### PR TITLE
ParamsArray type

### PR DESCRIPTION
**Pull request by alyst**
This PR is another subset of #193 changes and adds `ParamsArrays` type to simplify RAM specification.

`ParamsArray{T,N}` object defines N-dimensional array, whose elements are either constants of type `T` or parameters (specified as indices in parameters array).
`materialize(pararr, parvals)`, `materialize!(arr, parrarr, parvals)` and `sparse_materialize(pararr, parvals)` are used to create/update existing arrays with the `parvals` parameter values according to the `pararr::ParamsArray` specification.

It also supports fast sparse gradients by the parameters vector via `sparse_gradient!()`.

For lower level iteration over parameter references in an array, one can use the `param_occurences(pararr, i)` method that returns the linear indices of *i*-th param occurences.

The PR switches `RAMMatrices` to use the `ParamsArray` for *A* and *S* matrices and *M* vector.

### CODE REVIEW

File: src/StructuralEquationModels.jl [hunk 1]
```diff
@@ -69,7 +70,7 @@ include("optimizer/optim.jl")
 include("optimizer/NLopt.jl")
 # helper functions
 include("additional_functions/helper.jl")
-include("additional_functions/parameters.jl")
+#include("additional_functions/parameters.jl")
```
Comment (Maximilian-Stefan-Ernst) [inline src/StructuralEquationModels.jl:73]:
```suggestion
```

Should we delete the file `parameters.jl` then?

File: src/additional_functions/params_array.jl [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+Base.size(arr::ParamsArray) = arr.size
+Base.size(arr::ParamsArray, i::Integer) = arr.size[i]
+
+Base.:(==)(a::ParamsArray, b::ParamsArray) = return eltype(a) == eltype(b) &&
+       size(a) == size(b) &&
+       a.constants == b.constants &&
+       a.param_ptr == b.param_ptr &&
+       a.linear_indices == b.linear_indices
+
+Base.hash(a::ParamsArray, h::UInt) = hash(
```
Comment (Maximilian-Stefan-Ernst) [inline src/additional_functions/params_array.jl:82]: What do we need that for?

**Comment by Maximilian-Stefan-Ernst** [comment]
Thanks a lot, great PR - I just had some small comments, otherwise, i would merge it.

File: src/StructuralEquationModels.jl (hunk 1 above)
Comment (alyst) [reply src/StructuralEquationModels.jl:73]: done

**Comment by alyst** [comment]
> Thanks a lot, great PR - I just had some small comments, otherwise, i would merge it.

Thank you for the review! I've addressed your comments.
Also, I have rebased the PR against the current *devel* branch to keep the commit history clean.
I have locally rechecked for formatting of the code in the PR -- that should be fine, the CI check complains about the other files/line.

File: src/additional_functions/params_array.jl (hunk 2 above)
Comment (Maximilian-Stefan-Ernst) [reply src/additional_functions/params_array.jl:82]: Alright, seems reasonable - let's keep it here then.

**Pull request merged**